### PR TITLE
add gc finalizer mechanism

### PR DESCRIPTION
**Pull request by xujuntwt95329**

### CODE REVIEW

File: core/iwasm/include/gc_export.h [hunk 1]
```diff
@@ -898,6 +900,30 @@ wasm_runtime_pop_local_object_ref(wasm_exec_env_t exec_env);
 WASM_RUNTIME_API_EXTERN void
 wasm_runtime_pop_local_object_refs(wasm_exec_env_t exec_env, uint32_t n);
 
+/**
+ * Set finalizer to the given object, if multiple finalizer is set to the same
+ * object, previous one will be cancelled
```
Comment (wenyongh) [inline core/iwasm/include/gc_export.h:905]: Had better `the previous one`

File: core/iwasm/include/gc_export.h [hunk 2]
```diff
@@ -135,6 +135,8 @@ typedef struct WASMArrayObject *wasm_array_obj_t;
 typedef struct WASMFuncObject *wasm_func_obj_t;
 typedef uintptr_t wasm_i31_obj_t;
 
+typedef void (*wasm_finalizer_t)(const wasm_obj_t obj, void *data);
```
Comment (wenyongh) [inline core/iwasm/include/gc_export.h:138]: How about change to `wasm_obj_finalizer_t`?

File: core/shared/mem-alloc/ems/ems_gc.h [hunk 3]
```diff
@@ -186,6 +199,33 @@ gc_alloc_wo(void *heap, gc_size_t size);
 
 void
 gc_free_wo(void *vheap, void *ptr);
+
+extra_info_node_t *
+search_extra_info_node(gc_handle_t handle, gc_object_t obj, gc_size_t *p_index);
+
+/**
+ * Set finalizer to the given object, if multiple finalizer is set to the same
+ * object, previous one will be cancelled
```
Comment (wenyongh) [inline core/shared/mem-alloc/ems/ems_gc.h:208]: Same as above, had better change `multiple`, and `the previous`

File: core/shared/mem-alloc/ems/ems_gc_internal.h [hunk 4]
```diff
@@ -311,6 +311,10 @@ typedef struct gc_heap_struct {
     gc_size_t gc_threshold_factor;
     gc_size_t total_gc_count;
     gc_size_t total_gc_time;
+    extra_info_node_t *extra_info_normal_nodes[EXTRA_INFO_NORMAL_NODE_CNT];
+    extra_info_node_t **extra_info_nodes;
```
Comment (wenyongh) [inline core/shared/mem-alloc/ems/ems_gc_internal.h:315]: Could you add some comments about these fields?

File: core/shared/mem-alloc/ems/ems_alloc.c [hunk 5]
```diff
@@ -948,3 +948,147 @@ gci_dump(gc_heap_t *heap)
...
+        LOG_ERROR("alloc a new extra info node failed");
+        return GC_FALSE;
+    }
+    memset(node, 0, sizeof(extra_info_node_t));
+
+    node->finalizer = cb;
+    node->obj = obj;
+    node->data = data;
+
+    if (!insert_extra_info_node(vheap, node)) {
```
Comment (wenyongh) [inline core/shared/mem-alloc/ems/ems_alloc.c:1064]: Should lock the heap when accessing vheap's fields?

File: core/shared/mem-alloc/ems/ems_alloc.c [hunk 6]
```diff
@@ -948,3 +948,147 @@ gci_dump(gc_heap_t *heap)
...
+    gct_vm_set_extra_info_flag(obj, true);
+    return GC_TRUE;
+}
+
+void
+gc_unset_finalizer(gc_handle_t handle, gc_object_t obj)
+{
+    gc_size_t index;
+    gc_heap_t *vheap = (gc_heap_t *)handle;
+    extra_info_node_t *node = search_extra_info_node(vheap, obj, &index);
```
Comment (wenyongh) [inline core/shared/mem-alloc/ems/ems_alloc.c:1078]: Should lock the heap when accessing vheap's fields?

File: core/shared/mem-alloc/ems/ems_alloc.c [hunk 7]
```diff
@@ -948,3 +948,147 @@ gci_dump(gc_heap_t *heap)
...
+        return;
+    }
+
+    BH_FREE(node);
+    bh_memmove_s(
+        vheap->extra_info_nodes + index,
+        (vheap->extra_info_node_capacity - index) * sizeof(extra_info_node_t *),
+        vheap->extra_info_nodes + index + 1,
+        (vheap->extra_info_node_cnt - index) * sizeof(extra_info_node_t *));
+    vheap->extra_info_node_cnt -= 1;
```
Comment (wenyongh) [inline core/shared/mem-alloc/ems/ems_alloc.c:1090]: Same as above.

Comment (xujuntwt95329) [reply core/shared/mem-alloc/ems/ems_alloc.c:1097]: Done, thanks!

File: core/shared/mem-alloc/ems/ems_gc.c [hunk 8]
```diff
@@ -99,6 +100,17 @@ sweep_instance_heap(gc_heap_t *heap)
             /* merge previous free areas with current one */
             if (!last)
                 last = cur;
+
+            if (ut == HMU_WO) {
+                /* Invoke registered finalizer */
+                gc_object_t cur_obj = hmu_to_obj(cur);
+                if (gct_vm_get_extra_info_flag(cur_obj)) {
+                    extra_info_node_t *node = gc_search_extra_info_node(
+                        (gc_handle_t)heap, cur_obj, NULL);
```
Comment (wenyongh) [inline core/shared/mem-alloc/ems/ems_gc.c:109]: Had better add `bh_assert(node)`

Comment (xujuntwt95329) [reply core/shared/mem-alloc/ems/ems_gc.c:109]: Done, thanks

**Review by wenyongh — APPROVED** [review]
LGTM

**Pull request merged**